# Demo notebook for analysing location data

## Introduction

GPS location data contain rich information about people's behavioral and mobility patterns. However, working with such data is a challenging task since there exists a lot of noise and missingness. Also, designing relevant features to gain knowledge about the mobility pattern of subjects is a crucial task. To address these problems, `niimpy` provides these main functions to clean, downsample, and extract features from GPS location data:

- `niimpy.preprocessing.location.filter_location`: removes low-quality location data points
- `niimpy.util.aggregate`: downsamples data points to reduce noise
- `niimpy.preprocessing.location.extract_features_location`: feature extraction from location data

In the following, we go through analysing a subset of location data provided in [StudentLife](https://studentlife.cs.dartmouth.edu/dataset.html) dataset.

## Read data

In [1]:
import niimpy
from niimpy import config
import niimpy.preprocessing.location as nilo
import warnings
warnings.filterwarnings("ignore")

In [2]:
data = niimpy.read_csv(config.GPS_PATH, tz='Europe/Helsinki')
data.shape

(9857, 6)

There are 9857 location datapoints with 6 columns in the dataset. Let us have a quick look at the data:

In [3]:
data.head()

time  double_latitude  double_longitude  \
2013-03-27 06:03:29+02:00  1364357009        43.706667        -72.289097   
2013-03-27 06:23:29+02:00  1364358209        43.706637        -72.289066   
2013-03-27 06:43:25+02:00  1364359405        43.706678        -72.289018   
2013-03-27 07:03:29+02:00  1364360609        43.706665        -72.289087   
2013-03-27 07:23:25+02:00  1364361805        43.706808        -72.289370   

                           double_speed     user                  datetime  
2013-03-27 06:03:29+02:00          0.00  gps_u01 2013-03-27 06:03:29+02:00  
2013-03-27 06:23:29+02:00          0.00  gps_u01 2013-03-27 06:23:29+02:00  
2013-03-27 06:43:25+02:00          0.25  gps_u01 2013-03-27 06:43:25+02:00  
2013-03-27 07:03:29+02:00          0.00  gps_u01 2013-03-27 07:03:29+02:00  
2013-03-27 07:23:25+02:00          0.00  gps_u01 2013-03-27 07:23:25+02:00

The necessary columns for further analysis are `double_latitude`, `double_longitude`, `double_speed`, and `user`. `user` refers to a unique identifier for a subject.

## Filter data

Three different methods for filtering low-quality data points are implemented in `niimpy`:

- `remove_disabled`: removes data points whose `disabled` column is `True`.
- `remove_network`: removes data points whose `provider` column is `network`. This method keeps only `gps`-derived data points.
- `remove_zeros`: removes data points close to the point \<lat=0, lon=0\>.

In [4]:
data = nilo.filter_location(data, remove_disabled=False, remove_network=False, remove_zeros=True)
data.shape

(9857, 6)

There is no such data points in this dataset; therefore the dataset does not change after this step and the number of datapoints remains the same.

## Downsample

Because GPS records are not always very accurate and they have random errors, it is a good practice to downsample or aggregate data points which are recorded in close time windows. In other words, all the records in the same time window are aggregated to form one GPS record associated to that time window. There are a few parameters to adjust the aggregation setting:

- `freq`: represents the length of time window. This parameter follows the formatting of the pandas [time offset aliases](https://pandas.pydata.org/docs/user_guide/timeseries.html#timeseries-offset-aliases) function. For example '5T' means 5 minute intervals.
- `method_numerical`: specifies how numerical columns should be aggregated. Options are 'mean', 'median', 'sum'.
- `method_categorical`: specifies how categorical columns should be aggregated. Options are 'first', 'mode' (most frequent), 'last'.

The aggregation is performed for each `user` (subject) separately.

In [5]:
binned_data = niimpy.util.aggregate(data, freq='5T', method_numerical='median')
binned_data = binned_data.reset_index(0).dropna()
binned_data.shape

(9755, 5)

After binning, the number of datapoints (bins) reduces to 9755.

## Feature extraction

Here is the list of features `niimpy` extracts from location data:

1. distance based features (`niimpy.preprocessing.location.location_distance_features`):

| Feature         | Description |
|--------------|-----|
| `dist_total` |  total distance a person traveled in meters |
| `variance`, `log_variance`      | variance is defined as sum of variance in latitudes and longitudes |
|`speed_average`, `speed_variance`, and `speed_max`| statistics of speed (m/s). Speed, if not given, can be calculated by dividing the distance between two consequitive bins by their time difference|
|`n_bins`|number of location bins that a user recorded in dataset|

2. Significant Place related features (`niimpy.preprocessing.location.location_significant_place_features`):
 
 | Feature         | Description |
|--------------|-----|
|`n_static`| number of static points. Static points are defined as bins whose speed is lower than a threshold|
|`n_moving`| number of moving points. Equivalent to `n_bins - n_static`|
|`n_home`| number of static bins which are close to the person's home. Home is defined the place most visited during nights. More formally, all the locations recorded during 12 Am and 6 AM are clusterd and the center of largest cluster is assumed to be home|
|`max_dist_home`| maximum distance from home|
|`n_sps`| number of significant places. All of the static bins are clusterd using DBSCAN algorithm. Each cluster represents a Signicant Place (SP) for a user|
|`n_rare`| number of rarely visited (referred as outliers in DBSCAN)|
|`n_transitions`| number of transitions between significant places|
|`n_top1`| `n_top2`, `n_top3`, `n_top4`, `n_top5`: number of bins in the top `N` cluster. In other words, `n_top1` shows the number of times the person has visited the most freqently visited place|
|`entropy`| `normalized_entropy`: entropy of time spent in clusters. Normalized entropy is the entropy divided by the number of clusters|




In [6]:
import warnings
warnings.filterwarnings('ignore', category=RuntimeWarning)

# extract all the available features
all_features = nilo.extract_features_location(binned_data)
all_features

n_sps  n_static  n_moving  n_rare  n_home  max_dist_home  \
user                                                                
gps_u00     20      4095       152     164    2149   6.392724e+06   
gps_u01      3      5385       123       2      22   9.854291e+03   

         n_transitions  n_top1  n_top2  n_top3  ...  n_top5   entropy  \
user                                            ...                     
gps_u00            356    2171    1218     166  ...      63  7.243337   
gps_u01             10    5372       6       5  ...       0  2.339372   

         normalized_entropy    dist_total  n_bins  speed_average  \
user                                                               
gps_u00            2.417885  1.863604e+07    4247       0.289073   
gps_u01            2.129388  2.560327e+05    5508       0.043402   

         speed_variance  speed_max    variance  log_variance  
user                                                          
gps_u00        6.343650      34.00  184.473320      5.217505  
gps_u01        0.389312      32.75    0.000011    -11.410376  

[2 rows x 21 columns]

In [7]:
# extract only distance realted featuers
feature_functions = {
    nilo.location_distance_features: {} # arguments
}
distance_features = nilo.extract_features_location(
    binned_data,
    feature_functions=feature_functions)
distance_features

dist_total  n_bins  speed_average  speed_variance  speed_max  \
user                                                                      
gps_u00  1.863604e+07    4247       0.289073        6.343650      34.00   
gps_u01  2.560327e+05    5508       0.043402        0.389312      32.75   

           variance  log_variance  
user                               
gps_u00  184.473320      5.217505  
gps_u01    0.000011    -11.410376

The 2 rows correspond to the 2 users present in the dataset. Each column represents a feature. For example user `gps_u00` has higher variance in speeds (`speed_variance`) and location variance (`variance`) compared to the user `gps_u01`.

## Implementing your own features

If you want to implement a customized feature you can do so with defining a function that accepts a dataframe and returns a dataframe or a series. The reterned object should be indexed by `user`s. Then, when calling `extract_features_location` function, you add the newly implemented function to the `feature_functions` argument. The default feature functions implemented in `niimpy` are in this variable:

In [8]:
nilo.ALL_FEATURE_FUNCTIONS

{<function niimpy.preprocessing.location.location_significant_place_features(df, speed_threshold=0.277, min_samples=5, eps=200)>: {},
 <function niimpy.preprocessing.location.location_distance_features(df)>: {}}

You can add your new function to the `nilo.ALL_FEATURE_FUNCTIONS` dictionary and call `extract_features_location` function. Or if you are interested in only extracting your desired feature you can pass a dictionary containing just that function, like here:

In [9]:
# customized function
def max_speed(df):
    grouped = df.groupby('user')
    return grouped['double_speed'].max()

customized_features = nilo.extract_features_location(
    binned_data,
    feature_functions={max_speed: {}}
)
customized_features

double_speed
user                 
gps_u00         34.00
gps_u01         32.75